<p><a name="sections"></a></p>

# Sections

- <a href="#numpy">Numpy Overview</a>
- <a href="#array">Ndarray</a>
 - <a href="#dtype">Data Type</a>
 - <a href="#create">Creating Data Array</a>
- <a href="#slice">Subscripting and Slicing</a>
 - <a href="#shape">Shape</a>
 - <a href="#operations">Operations</a>
 - <a href="#matrix">Matrix and Linear Algebra</a>
- <a href="#sample">Random Sampling</a>
 - <a href="#birthday">Random Sampling: Birthday Problems</a>
- <a href="#scipy">Scipy Overview</a>
 - <a href="#statistics">Statistical Functions</a>
 - <a href="#hypothesis">Hypothesis Test</a>
- <a href="#sol">Solutions</a>

<p><a name="numpy"></a></p>
## NumPy Overview

- NumPy is the fundamental package for scientific computing with Python. 
 - [Primitive data types](http://docs.scipy.org/doc/numpy/user/basics.types.html) are represented in certain ways to facilitate easy representation and manipulation. Common examples in linear algebra are **vectors** and **matrices**. Numpy provides excellent tools to work with these numerical types. 
 - Numpy provides a comprehensive set of functions and modules for scientific computing, such as linear algebra operations, random number generation, Fourier transform, etc.
 - For full documentation, go to:  http://docs.scipy.org/doc/.
 - To use NumPy, import the module as follows:

In [1]:
import numpy as np

- Whenever you need to call a function from Numpy, say `matrix()`, you could run:
```
np.matrix( arguments )
```

<p><a name="array"></a></p>
## Ndarray

<p><a name="dtype"></a></p>
### Data Type
- NumPy provides an N-dimensional array type, the ndarray. It can be described as a shaped collection of elements of the same data type. 
- The following code constructs a simple 1-D array using the function `array()`.

In [4]:
dir(np.pi)

['__abs__',
 '__add__',
 '__bool__',
 '__class__',
 '__delattr__',
 '__dir__',
 '__divmod__',
 '__doc__',
 '__eq__',
 '__float__',
 '__floordiv__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getformat__',
 '__getnewargs__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__int__',
 '__le__',
 '__lt__',
 '__mod__',
 '__mul__',
 '__ne__',
 '__neg__',
 '__new__',
 '__pos__',
 '__pow__',
 '__radd__',
 '__rdivmod__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__rfloordiv__',
 '__rmod__',
 '__rmul__',
 '__round__',
 '__rpow__',
 '__rsub__',
 '__rtruediv__',
 '__set_format__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 '__truediv__',
 '__trunc__',
 'as_integer_ratio',
 'conjugate',
 'fromhex',
 'hex',
 'imag',
 'is_integer',
 'real']

In [5]:
help(np.pi)

Help on float object:

class float(object)
 |  float(x=0, /)
 |  
 |  Convert a string or number to a floating point number, if possible.
 |  
 |  Methods defined here:
 |  
 |  __abs__(self, /)
 |      abs(self)
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __bool__(self, /)
 |      self != 0
 |  
 |  __divmod__(self, value, /)
 |      Return divmod(self, value).
 |  
 |  __eq__(self, value, /)
 |      Return self==value.
 |  
 |  __float__(self, /)
 |      float(self)
 |  
 |  __floordiv__(self, value, /)
 |      Return self//value.
 |  
 |  __format__(self, format_spec, /)
 |      Formats the float according to format_spec.
 |  
 |  __ge__(self, value, /)
 |      Return self>=value.
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __getnewargs__(self, /)
 |  
 |  __gt__(self, value, /)
 |      Return self>value.
 |  
 |  __hash__(self, /)
 |      Return hash(self).
 |  
 |  __int__(self, /)
 |      int(self)
 |  
 |  __le__

In [2]:
my_ary=np.array([1,2,3])

- We see that `array()` takes a list as an argument. Then `my_ary` is a ndarray with three elements ordered exactly the same as the list inputted.

In [3]:
print(type(my_ary))
print(my_ary)

<class 'numpy.ndarray'>
[1 2 3]


- Each element in my_ary can be accessed by an integer index. Details of indexing will be provided later.

In [4]:
print(my_ary[0])
print(my_ary[1])

1
2


- The type ndarray is flexible because it supports great variety of primitive data types: subject to the **homogeneity condition**, which means every element in ndarray have the same type. 
- The type of the elements can be accessed as an attribute of an ndarray.

In [5]:
my_ary.dtype

dtype('int32')

- When the elements are of string types, it print the dtype automatically. 
- The letter `U` stands for unicode and the number followed represents the largest number of letters among all the elements. 

In [6]:
np.array(['a', 'ab'])

array(['a', 'ab'], dtype='<U2')

- If the list inputted to the function `array()` is of mixed types, Numpy will upcast all the elements to a common superclass:

In [7]:
print(np.array([1.0, 2, 3]))
print(np.array(['a', 2, 3]))

[1. 2. 3.]
['a' '2' '3']


- The most general type for numpy arrays is the generic type `object`.

In [8]:
np.array([[],0,True,'string'])

array([list([]), 0, True, 'string'], dtype=object)

<p><a name="create"></a></p>
### Creating Data Array

- So far a ndarray has been created by passing a list to the numpy function `array()`. 
- This method can be generalized to create multidimensional arrays by passing a nested list to the `array()` function.

In [9]:
simple_lst = [1,2]
nested_lst = [[1,2], [3,4]]
multi_ary = np.array(nested_lst)
multi_ary

array([[1, 2],
       [3, 4]])

- Recall that a nested list has lists as elements. The multidimensional ndarray is designed to effectively represent the two dimensional structure of mathematical matrices. 
- In this case, each inner list of nest_lst is like a row in a matrix.
- A three dimensional array may be constructed by passing a list of two dimensional arrays to `array()`. In the scope of the course most emphasis is on gaining proficiency using one and two dimensional arrays.

In [10]:
nested_lst2 = [[5,6], [7,8]]
multi_ary2 = np.array(nested_lst2)
higher_dim = np.array([multi_ary, multi_ary2])
higher_dim

array([[[1, 2],
        [3, 4]],

       [[5, 6],
        [7, 8]]])

- Numpy provides several functions for constructing arrays without having to specify the value of each element. Some are shown of them below.
- One such function is `arange()`. It operates much like the native Python function `range()`.

In [11]:
np.arange(10)

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

- Passing a float to `arange()` returns a 1-D array with elements from zero up to the given float (exclusive).

In [12]:
np.arange(9.5)

array([0., 1., 2., 3., 4., 5., 6., 7., 8., 9.])

- An alternate starting number can be specified. The first argument below indicates the array starts from 2 and the second argument indicates it goes to 10 (exclusive). Therefore is includes the integers 2 through 9.

In [13]:
np.arange(2, 10)

array([2, 3, 4, 5, 6, 7, 8, 9])

- By default arange increments by 1. A third argument can be included to change the incrementation. 
- In the example below the start is 2 and the increment is 3. The following values are therefore, 2 + 3 = 5, then 5 + 3 = 8. Since the next one, 8 + 3 = 11, is greater than the upper bound 9, the sequence stops at 8.

In [14]:
np.arange(2,10,3)

array([2, 5, 8])

- A similar numpy method is `linspace()`. It takes the start (first argument), the end (second argument) and the number of elements (last argument).
- Without the last argument, the default output length equals to 50.

In [15]:
np.linspace(0,10,51)

array([ 0. ,  0.2,  0.4,  0.6,  0.8,  1. ,  1.2,  1.4,  1.6,  1.8,  2. ,
        2.2,  2.4,  2.6,  2.8,  3. ,  3.2,  3.4,  3.6,  3.8,  4. ,  4.2,
        4.4,  4.6,  4.8,  5. ,  5.2,  5.4,  5.6,  5.8,  6. ,  6.2,  6.4,
        6.6,  6.8,  7. ,  7.2,  7.4,  7.6,  7.8,  8. ,  8.2,  8.4,  8.6,
        8.8,  9. ,  9.2,  9.4,  9.6,  9.8, 10. ])

In [16]:
np.linspace(0,10)

array([ 0.        ,  0.20408163,  0.40816327,  0.6122449 ,  0.81632653,
        1.02040816,  1.2244898 ,  1.42857143,  1.63265306,  1.83673469,
        2.04081633,  2.24489796,  2.44897959,  2.65306122,  2.85714286,
        3.06122449,  3.26530612,  3.46938776,  3.67346939,  3.87755102,
        4.08163265,  4.28571429,  4.48979592,  4.69387755,  4.89795918,
        5.10204082,  5.30612245,  5.51020408,  5.71428571,  5.91836735,
        6.12244898,  6.32653061,  6.53061224,  6.73469388,  6.93877551,
        7.14285714,  7.34693878,  7.55102041,  7.75510204,  7.95918367,
        8.16326531,  8.36734694,  8.57142857,  8.7755102 ,  8.97959184,
        9.18367347,  9.3877551 ,  9.59183673,  9.79591837, 10.        ])

**Exercise 1**

- Which of the following two arrays has all integer (this refer to the value, not data type) entries? 
 - `np.linspace(0, 9, 3)`
 - `np.linspace(0, 9, 4)`

- As `array()`, we can change the datatype by specifying dtype. Run the following code, what do you find?
 - `np.linspace(0, 9, 4, dtype ='Int64')`
 - `np.linspace(0, 9, 3, dtype ='Int64')`

In [ ]:
### Your code here


**Spawning Arrays**

- To construct an array of identical values, the function `ones()` is useful.  
- It returns an array filled with values of one.  The argument determines the number of elements.

In [17]:
np.ones(3)

array([1., 1., 1.])

- We can construct all the constant array by multiply the desired constant to a array produced as above. 
- The details of array operation will be discussed, but multiplying a number to an array results in every entry in the array multiplied by the constant.

In [18]:
import math
math.pi *np.ones(5)

array([3.14159265, 3.14159265, 3.14159265, 3.14159265, 3.14159265])

- Numpy also provides a function `zeros()`, which works in the same way as the function `ones()`:

In [19]:
np.zeros(5)

array([0., 0., 0., 0., 0.])

<p><a name="slice"></a></p>
## Subscripting and Slicing

- To access a particular element in an array, indices are used.  This is also known as subscripting.  
- Array entries are indexed by integers.

In [20]:
x = np.arange(1, 11)
x

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

- We may print out an arbitrary entry. For example, the 3rd element in x:

In [21]:
x[2]  # Python index from 0

3

- We may then plug it into a function:

In [22]:
3**(x[2])

27

- We may update an element with indexes:

In [23]:
x[2] = 100
x

array([  1,   2, 100,   4,   5,   6,   7,   8,   9,  10])

- Negative indexes select the element by the opposite order.

In [24]:
print(x[-1])
print(x[-2]) 
print(x[-10])

10
9
1


**Exercise 2**

- Initialize your x again with:
```
x = np.arange(1, 11)
```
- Run `x[2]=3.0`. Which entry of the array x will be updated?
- Run `x[2]=3.1`. Which entry of the array x will be updated?
- Run `x.astype(float)` to change the type to float, then update the third element to 3.1.

In [25]:
x = np.arange(1, 11)
x

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

In [26]:
#### Your code here


- It is often necessary to subscript in a higher dimensional array. Start by constructing one below:

In [27]:
high_x = np.array([[1,2,3],[4,5,6]])
high_x

array([[1, 2, 3],
       [4, 5, 6]])

As we mentioned before, 2-dimensional ndarray can be thought as a nested list. Therefore, to access the number 2, it is the element 1 in the list 0:

In [28]:
high_x[0][1]

2

The whole row 0 may be selected as follows:

In [29]:
high_x[0]

array([1, 2, 3])

- A 2-dimensional array can be also taken for a matrix.

\begin{pmatrix}
1 &2  &3 \\ 
4 &5  &6 
\end{pmatrix}

- Therefore as matrices, a 2-dimensional array can be indexed a pair of integers -- row indexes and columns indexes. Conventionally, the first index is for row, the second for column.


In [30]:
print(high_x[0,0]) 
print(high_x[0,1])
print(high_x[0,2])
print(high_x[1,0])
print(high_x[1,1])
print(high_x[1,2])

1
2
3
4
5
6


- The first two rows may be accessed using slicing, as follows:

In [31]:
high_x[:2]

array([[1, 2, 3],
       [4, 5, 6]])

- A combination of slicing and indexing can be used to access the second column of the first two rows.

In [32]:
high_x[:2,1]

array([2, 5])

<p><a name="shape"></a></p>
### Shape

- Again as a matrix, the shape of an array can be denoted by the number of rows and columns it has. And again, first rows and then columns. For example, consider the following 2 by 3 matrix.

\begin{pmatrix}
1 &2  &3 \\ 
4 &5  &6 
\end{pmatrix}

- For an array, the shape (or dimension in more matrix convention) can be seen by calling the `.shape` attribute.

In [33]:
high_x = np.array([[1,2,3],[4,5,6]])
high_x.shape

(2, 3)

- When constructing arrays from functions like `np.ones()`, a tuple or list may be passed as an argument to determine the shape of the resulting array.

In [34]:
np.ones([2,3])

array([[1., 1., 1.],
       [1., 1., 1.]])

- The function `zero()` works in the same fashion:

In [35]:
np.zeros([3,2])

array([[0., 0.],
       [0., 0.],
       [0., 0.]])

- The shape of an array can be changed using the method `reshape()`, as follows:

In [36]:
x = np.arange(8)
x.reshape([2,4])

array([[0, 1, 2, 3],
       [4, 5, 6, 7]])

- Reshaping can also be achieved by changing the `shape` attribute directly.

In [37]:
x = np.arange(8)
x.shape = (2,4)
x

array([[0, 1, 2, 3],
       [4, 5, 6, 7]])

**Exercise 3**

- Create a 4 by 5 five matrix with all the  entries equal to 8.
- Create an array corresponding to the matrix below:
\begin{pmatrix}
9 &8  &7 \\ 
6 &5  &4 
\end{pmatrix}

In [38]:
#### Your code here


<p><a name="operations"></a></p>
### Operations

**Arithmetic Operations**

- Numpy allows mathematical operations to be performed on arrays with ease. 
- Addition of more than one array works as expected, as long as they share the same length:

In [39]:
x = np.array([1,2])
y = np.array([3,4])
z = np.array([5,6])
x + y + z

array([ 9, 12])

- Addition is an example of **pointwise operations**. This means each element in one array operates on it's corresponding element in the other array.
- Here is another example:

In [40]:
y*z       # Same rule applies to multiplication

array([15, 24])

All the arithmetic operators can work pointwisely

In [41]:
x**y

array([ 1, 16], dtype=int32)

In [42]:
y/x 

array([3., 2.])

In [43]:
y-x

array([2, 2])

In [44]:
z%y 

array([2, 2], dtype=int32)

**Broadcasting** 
- Broadcasting describes the situation where a unary operator, or binary operator paired with a scalar, is distributed to every element in an array.  Here are examples:

In [45]:
x+3

array([4, 5])

- All the arithmetic operators can be broadcasted similarly

In [46]:
x**2

array([1, 4], dtype=int32)

In [47]:
x/2

array([0.5, 1. ])

In [48]:
y-3

array([0, 1])

In [49]:
z%3   

array([2, 0], dtype=int32)

**Comparison Operations**

- Comparison operators generate Boolean arrays. As binary arithmetic operators, they also follow the broadcasting rule when paired with a scalar.

In [50]:
x>1

array([False,  True])

- All the comparison operators can be broadcasted similarly.

In [51]:
x<1

array([False, False])

In [52]:
y<=3

array([ True, False])

In [53]:
z>=6

array([False,  True])

In [54]:
z==5

array([ True, False])

- Comparison operators acting between arrays work **pointwise**.  
- Each element in the left array is compared to the corresponding element in the right array.  The two arrays must therefore have the same shape.

In [55]:
x>y

array([False, False])

In [56]:
y<=z

array([ True,  True])

In [57]:
z>=x

array([ True,  True])

In [58]:
z==x

array([False, False])

**Logical Operators**

- The logical operators `and, or` (`&,|`) work pointwisely when acting between two boolean arrays of thew same shape.

In [59]:
b_1 = np.array([True, True, True])
b_2 = np.array([True, True, False])
b_3 = np.array([True, False, False])
b_4 = np.array([False, False, False])

In [60]:
#logical and 
print(b_1 & b_2)
print(np.logical_and(b_1,b_2))

[ True  True False]
[ True  True False]


In [61]:
#logical or
print(b_2 | b_4)
print(np.logical_or(b_1,b_2))

[ True  True False]
[ True  True  True]


- The logical operator `not`  is unary.  It only has one argument.  When acting on a boolean array it is broadcast to all the elements.

In [62]:
#logical not
print(~b_3)
print(np.logical_not(b_3))

[False  True  True]
[False  True  True]


**Aggregating Boolean Arrays**

- The `all` operator outputs the `logical_and` between all the elements of a boolean array.  
- The `any` operator outputs the `logical_or` between all the elements of a boolean array.
- These types of operations are known as aggregating.

In [63]:
print(b_1.all())
print(b_2.all())

True
False


In [64]:
print(b_3.any())
print(b_4.any())

True
False


**Fancy Indexing**

- We often need to filter an array according to some condition. This can be done by two steps.

 - Generate Boolean array

In [65]:
x

array([1, 2])

In [66]:
x==1

array([ True, False])

- Slice the array with the Boolean array

In [67]:
x[np.array([ True, False])]

array([1])

- The two steps can be actually combined

In [68]:
x[x==1]

array([1])

- Fancy indexing can be applied to arrays of any dimension. 
- Slicing with a boolean array, as is done in fancy indexing, drops the shape of the array.  A one dimensional array is returned.

In [71]:
high_x = np.array([[1,2,3],[4,5,6]])
high_x[high_x>=3]

array([3, 4, 5, 6])

- However, assignment with fancy indexing doesn’t change the shape.

In [72]:
high_x[high_x>=3]=10
high_x

array([[ 1,  2, 10],
       [10, 10, 10]])

** Exercise 4**

- Run the code below to create arrays.

```
ary_1 = np.ones([3,2])
ary_2 = np.arange(1,7).reshape(3,2)
```
- Sum up the arrays and select the elements from the result array that are greater than 2 and less than 6

In [73]:
ary_1 = np.ones([3,2])
ary_2 = np.arange(1,7).reshape(3,2)

In [74]:
#### Your code here


<p><a name="matrix"></a></p>
### Matrix and Linear Algebra

**Matrix type**

- Numpy has two similar object types, arrays and matrices. A matrix is more narrow in scope as it represents a 2-D structure only.  
- Numpy provides a function `matrix()` that returns the matrix object type `numpy.matrixlib.defmatrix.matrix`.  The `matrix()` function can be passed a list or an array.

In [75]:
x = np.matrix([3,2])
print(type(x))
print(x.shape)

<class 'numpy.matrixlib.defmatrix.matrix'>
(1, 2)


- Once an array is converted to a matrix, it is taken for a **2-dimensional object**. 
- The difference lies in the fact that, we can access the second entry of a numpy array by:

In [76]:
y=np.array([3,4])
print(y[1])
print(y.shape)

4
(2,)


- However, if we change y into a matrix, it no longer works.
- The expression `y[1]` now indicates the second row of `y`, which it does not have. 

In [77]:
y = np.matrix(y)
y[1]

IndexError: index 1 is out of bounds for axis 0 with size 1

- Since `y` is now taken for 2-dimensional (Notice the extra bracket):

In [78]:
y

matrix([[3, 4]])

- Note the shape compared to the shape of the original array.

In [79]:
y.shape

(1, 2)

- To access the element 4, we need to select the second element in the first row by:

In [80]:
y[0, 1]

4

- Multiple inner lists are considered multiple rows:

In [81]:
z = np.matrix([[3],[4]])
z

matrix([[3],
        [4]])

**Matrix Multiplication**

- Another important difference between arrays and matrices is the multiplication. 
- While the comparison and the other arithmetic operations still works pointwisely, multiplication (and therefore raising power) does not. 
- Run the code below and look at the error message.

In [82]:
x

matrix([[3, 2]])

In [83]:
y

matrix([[3, 4]])

In [84]:
try:
    x*y     # * is matrix multiplication
except Exception as e:
    print(e)

shapes (1,2) and (1,2) not aligned: 2 (dim 1) != 1 (dim 0)


- The type matrix is for linear algebra, in which the matrix multiplication is always **“a row times a column”**:

$$
(1,2,3) \times \left(
\begin{array}{c}
1\\
2\\
3\\
\end{array}
\right)
= (1+4+9) = 14
$$

- The above multiplication is known at the "dot product." Both `x` and `y` have been transformed from ndarray into matrix type, which is taken for a row vector (matrix with 1 row, if you want). 
- To change the later to a column vector (matrix with 1 column), we may use the method `.T`. 

In [85]:
print(y)
print(y.shape)

[[3 4]]
(1, 2)


In [86]:
print(y.T)
print(y.T.shape)

[[3]
 [4]]
(2, 1)


In [87]:
print(x.shape)

(1, 2)


The multiplication can be performed as follows.  This operation is known as the inner product.  It yields a single value.

In [88]:
xy1=x*y.T
print(xy1)
print(xy1.shape)

[[17]]
(1, 1)


- Note how the inner values of the shapes must match.
```
(1,2)*(2,1) ----> (1,1)
```

- The multiplication may be performed in an alternate fashion.
```
(2,1)*(1,2) ----> (2,2)
```
- Note the inner values of the shape still match but a 2x2 matrix is returned.  This is known as the outer product.

In [89]:
xy2=x.T*y
print(xy2)
print(xy2.shape)

[[ 9 12]
 [ 6  8]]
(2, 2)


- Matrix multiplication may also be performed using arrays and the object method `dot`.  
- It works in much the same way `*` works for matrices but is applicable to higher dimensional arrays. This makes it more powerful.  
- The shapes must fit together in the same way as matrix multiplication.  The second dimension in the left array must match the first dimension of the right array.
```
(m,n) dot (n,k) ----> (m,k)
```

In [90]:
x=np.array(x)
y=np.array(y)
xy1=x.dot(y.T)
xy2=x.T.dot(y)
print('inner product\n',xy1)
print('-'*13)
print('outer product\n',xy2)

inner product
 [[17]]
-------------
outer product
 [[ 9 12]
 [ 6  8]]


**Identity Matrix**

- The identity matrix is a square matrix with all ones in the diagonal elements and other elements zero. 
- An identity array can be created by specifying the dimension in the function `eye()`:

In [91]:
np.eye(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

- We then change it into a matrix type.

In [92]:
Id = np.matrix(np.eye(3))

- Multiplication of another matrix by the identity matrix leaves the other unchanged. 
- In this way it acts like the number one for standard multiplication of scalars; thus the name.  

In [93]:
my_mat = np.matrix(range(9))
my_mat = my_mat.reshape([3,3])
my_mat

matrix([[0, 1, 2],
        [3, 4, 5],
        [6, 7, 8]])

In [94]:
#left multiplication by the identity matrix
Id*my_mat

matrix([[0., 1., 2.],
        [3., 4., 5.],
        [6., 7., 8.]])

In [95]:
#right multiplication by the identity matrix
my_mat*Id

matrix([[0., 1., 2.],
        [3., 4., 5.],
        [6., 7., 8.]])

**Inverse**

- The identity matrix is analogous to number one. One is referred to as the multiplicative identity, since any number multiplied by one remains unchanged. 
- A nonzero number has a multiplicative inverse, or reciprocal.  The product of a number and its multiplicative inverse is the multiplicative identity, one. 

**For real scalars:**

- nonzero number: n
- multiplicative identity: 1
- multiplicative inverse: $\frac{1}{n}$
$$n \times 1 = n$$

$$n \times \frac{1}{n}=1$$

- The analogy for a square matrix is called an inverse matrix. The product of a matrix and its inverse matrix (in either order) is the identity matrix. 

**For real-valued square matrices:**

- nonsingular matrix: $M$
- identity matrix: $\textbf{1}$
- inverse: $M^{-1}$
$$M \cdot \textbf{1} = M$$

$$M \cdot M^{-1}=\textbf{1}$$

$$M^{-1} \cdot M=\textbf{1}$$

In [96]:
m1 = np.matrix([[1,2],[0,1]])
m1

matrix([[1, 2],
        [0, 1]])

In [97]:
m2 = np.matrix([[1,-2],[0,1]])
m2

matrix([[ 1, -2],
        [ 0,  1]])

- Computing the product of m1 and m2 (in both orders), yields the identity.  Therefore the inverse of m1 is m2 and vice versa.

In [98]:
print(m1*m2,'\n\n',m2*m1)

[[1 0]
 [0 1]] 

 [[1 0]
 [0 1]]


- If a matrix has an inverse, it is unique. Also note, the inverse of the inverse is the original matrix.

- For a numpy matrix (with an inverse), there is a method returning the inverse matrix:

In [99]:
print(m1)
print('\n')
print(m1.I)
print('\n')
print(m1.I.I)

[[1 2]
 [0 1]]


[[ 1. -2.]
 [ 0.  1.]]


[[1. 2.]
 [0. 1.]]


**Exercise 5**

- How do you find a tuple of `x`, `y` and `z` satisfying all the equations below?

$$
\begin{eqnarray}
3x +2y -z &= 1\\
2x -2y +4z &=-2\\
-x +\frac{1}{2}y- z&= 0
\end{eqnarray}
$$

Observe that the equation can be written as:

$$
\begin{pmatrix}
3 &2  &-1 \\ 
2 &-2  &4 \\
-1&\frac{1}{2}&-1
\end{pmatrix}
\times
\begin{pmatrix}
x\\y\\z
\end{pmatrix}
=
\begin{pmatrix}
1\\-2\\0
\end{pmatrix}
$$

$$\begin{pmatrix}
x\\y\\z
\end{pmatrix}=
\begin{pmatrix}
3 &2  &-1 \\ 
2 &-2  &4 \\
-1&\frac{1}{2}&-1
\end{pmatrix}^{-1} \times
\begin{pmatrix}
1\\-2\\0
\end{pmatrix}
$$


From here, can we use **inverse** to solve for x, y and z?

In [100]:
#### Your code here


<p><a name="sample"></a></p>
### Random Sampling

- We often want to generate array randomly for simulation or testing purpose. 
- Both Numpy and SciPy provide functions for random sampling.

**List of Functions for Random Sampling in Numpy**

- The `numpy.random` submodule provides random number generator. We list some below.

 - `randn(d0, d1, ..., dn)`: Each entry is from the “standard normal” distribution.

 - `rand(d0, d1, ..., dn)`: Random array with a given shape where each element is from Uniform [0, 1].

 - `randint(low, high, size)`: Random integers ranging from low (inclusive) to high (exclusive).

 - `choice(a[, size, replace, p])`: Generates a random sample from a given 1-D array. By default, `replace=True`.

<p><a name="birthday"></a></p>
#### Random Sampling: Birthday Problems

- If we randomly choose 20 people, what is the probability that some of them share the same birth day? For people who are familiar with probability, you might want to compute the probability that nobody sharing the same birthday and subtract that from 1. You are right that this is doable.
- We will approach a solution with another interesting method -- simulation. We know that there would be 366 different possible birthday, including Feb. 29. Assuming the probability for each date to be picked is the same (obviously this is not valid, we assume it for simplicity)
- To simulate the situation, we may use the `choice()` function from the `numpy.random` submodule.

In [101]:
np.random.seed(1)
np.random.choice(range(366), size=20, replace=True)

array([ 37, 235,  72, 255, 203, 133, 335, 144, 129,  71, 237, 281, 178,
       276, 254, 357, 252, 156,  50,  68])

In [102]:
np.random.choice(range(366), size=20, replace=True)

array([215, 241, 352,  86, 141,   7, 319, 317,  22, 313,   1, 316, 209,
       264, 216, 141, 115, 121,  30,  71])

- The first simulation has no duplicated dates and the second does have 141 twice. If we only simulate twice, we might think that the desired probability is 0.5!

- However, this is obviously very biased! What we want to do is to simulate many times and calculate the probability according to that. Below are the code for simulation.

In [103]:
num_people=20
num_simu=int(1e4)
Bool = np.zeros(num_simu)
for i in range(num_simu):
    test = np.random.choice(range(366), size=num_people, replace=True)
    Bool[i] = (len(set(test))!=num_people)
np.mean(Bool)

0.4137

<p><a name="scipy"></a></p>
## Scipy Overview

- SciPy is a package for scientific computing and technical computing with Python. 
 - SciPy builds on the NumPy array object and is part of the NumPy stack which includes tools like Matplotlib and pandas.
 - SciPy contains modules for stats (which we will focus on), integration, optimize, linear algebra, interpolation, special functions, FFT, signal, image processing, ODE solvers and more.
 - For full documentation, go to:  http://docs.scipy.org/doc/.

**Sample data set**

- Before beginning the class in SciPy, we first import the modules. We also import datasets from `sklearn` module to access the sample data set -- iris.

In [104]:
import numpy as np
from scipy import stats

from sklearn import datasets
iris_raw = datasets.load_iris()

- The package sklearn (scikit-learn) is for machine learning. It contains a well known data set, iris, containing measurements on different species of flowers called irises.  
- The measured features are the length and width of the petals and sepals. The species of each flower is also known.  
- In the following exploration the width and length of sepal and petal are used as numeric features.

In [105]:
type(iris_raw)

sklearn.datasets.base.Bunch

In [108]:
iris_raw.keys()

dict_keys(['feature_names', 'DESCR', 'target', 'data', 'target_names'])

In [107]:
print(iris_raw['DESCR'])

Iris Plants Database

Notes
-----
Data Set Characteristics:
    :Number of Instances: 150 (50 in each of three classes)
    :Number of Attributes: 4 numeric, predictive attributes and the class
    :Attribute Information:
        - sepal length in cm
        - sepal width in cm
        - petal length in cm
        - petal width in cm
        - class:
                - Iris-Setosa
                - Iris-Versicolour
                - Iris-Virginica
    :Summary Statistics:

    ============== ==== ==== ======= ===== ====================
                    Min  Max   Mean    SD   Class Correlation
    ============== ==== ==== ======= ===== ====================
    sepal length:   4.3  7.9   5.84   0.83    0.7826
    sepal width:    2.0  4.4   3.05   0.43   -0.4194
    petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)
    petal width:    0.1  2.5   1.20  0.76     0.9565  (high!)
    ============== ==== ==== ======= ===== ====================

    :Missing Attribute Values: None
  

- The iris data is recorded in `iris_raw.data` as a 2-dimensional array. There are 150 rows corresponding to 150 instances (flowers), 4 columns correspond to 4 different numerical features.  
- There are 50 rows of numerical data for each of 3 species of iris.

In [109]:
iris = iris_raw['data']
iris[0:8,:]

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2]])

- Extract the following three feature column arrays, as follows:

In [110]:
sepal_len = iris[:,0]
sepal_wid = iris[:,1]
petal_len = iris[:,2]

- The labels of iris data is recorded in `target` as an array:

In [111]:
iris_raw['target']

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

- We assign the target array to a variable called “species”:

In [112]:
species = iris_raw['target']

- Though the values are integers, they are represent the species.  The actual names are available in `target_names`. 

In [113]:
iris_raw['target_names']

array(['setosa', 'versicolor', 'virginica'], dtype='<U10')

<p><a name="statistics"></a></p>
### Statistical Functions

- A descriptive statistical digest is useful when confronting new data, for both categorical and numerical features.
- The stats function `itemfreq()` returns the frequency of each class in a categorical feature.

In [114]:
stats.itemfreq(species)

array([[ 0, 50],
       [ 1, 50],
       [ 2, 50]], dtype=int64)

- Numerical features are summarized by the stats function `describe()`:

In [115]:
stats.describe(sepal_len)

DescribeResult(nobs=150, minmax=(4.3, 7.9), mean=5.843333333333334, variance=0.6856935123042507, skewness=0.3117530585022963, kurtosis=-0.5735679489249765)

- The function returns axescriptive object `DescribeResult`. To access each of item, use the item name or integer index:

In [116]:
stats.describe(sepal_len)[0]

150

In [117]:
stats.describe(sepal_len).minmax

(4.3, 7.9)

More on statistical functions can be found in the link.
- http://docs.scipy.org/doc/scipy/reference/stats.html

<p><a name="hypothesis"></a></p>
### Hypothesis Test

- Forming a testable hypothesis and deriving the p-value (probability) that the hypothesis is valid, is an effective technique for drawing statistical conclusions from observations of populations. 
- The null hypothesis states there is no statistical significance between specified populations. 
- If the p-value under the null hypothesis is low it means there is a low probability there is no statistical significance between specified populations.  In other words, rejection of the null hypothesis is an affirmation that there is statistical significance between specified populations.  
- The contrary to the null hypothesis is called the alternative hypothesis.  It is retained if the null hypothesis is rejected.
- The stats module provides functions to perform a variety of hypothesis tests. The most common ones are demonstrated below.

**One Sample t-test**

- The one-sample t-test is used to determine whether a sample belongs to a population with a known mean.  
- For example, to know how likely that the sample array `petal_len` belongs to a population of mean 10,  use `ttest_1samp()` as follows:

In [118]:
stats.ttest_1samp(petal_len, 10)

Ttest_1sampResult(statistic=-43.323240335498866, pvalue=2.4562011895365267e-86)

- In this case, the p-value is extremely small.  This indicates it is **unlikely** the sample was taken from a population of with mean petal length 10.

**Two Sample t-test**

- The two-sample t-test is used to test the hypothesis that the populations from which the two samples were taken have the same mean.  
- For example, the arrays `sepal_len` and `petal_len` were sampled from populations of septal and petal length.  
- To know how likely it is the population means are equal, use `ttest_ind()` as follows:

In [119]:
stats.ttest_ind(sepal_len, petal_len)

Ttest_indResult(statistic=13.099504494510061, pvalue=2.8297338637366177e-31)

- In this case, the p-value is extremely small.  This indicates the populations from which `sepal_len` and `petal_len` are sampled are **unlikely** to have the same mean.  In other words, the difference in sepal and petal length are statistically different.

**ANOVA**

- The two-sample t-test checks whether the populations from which two samples are taken have a statistically significant difference in mean.  Analysis of Variance extends this technique for any number of samples.  
- Consider the three samples of sepal length, sepal width and petal length.  Analysis of Variance, or ANOVA, is performed on these three samples using the `stats` function `f_oneway`, as follows:

In [120]:
stats.f_oneway(sepal_len, sepal_wid, petal_len)

F_onewayResult(statistic=237.45598625019218, pvalue=5.440492586967051e-71)

- The function `f_oneway` is used to implement one-way analysis of variance. In this particular case, since the p-value is very small, it is **unlikely** that all populations, from which the samples were taken, have the same mean.  In other words, the difference in these features is statistically significant.

- More on hypothesis tests can be found from the page statistical functions, which we have seen before.
 - http://docs.scipy.org/doc/scipy/reference/stats.html

**Exercise 6**

- Consider the following fancy indexing used to extract the sepal length values for the iris species setosa.
```
sepal_len[iris_raw['target']==0]
```

- Use this technique to extract the sepal lengths for the three species of iris (setosa, versicolor, virginica) and assign them to variables of your choosing.

- Next use ANOVA to determine if the three species have a statistical difference in their sepal lengths.

In [ ]:
# Your code here


<p><a name="sol"></a></p>
# Solutions

**Exercise 1**

In [ ]:
print(np.linspace(0, 9, 3))
print(np.linspace(0, 9, 4))
print(np.linspace(0, 9, 3, dtype='Int64'))
print(np.linspace(0, 9, 4, dtype='Int64'))

**Exercise 2**

In [ ]:
x=x.astype(float)
x[2]=3.1
x

**Exercise 3**

In [ ]:
np.ones([4,5])*8.0
np.arange(9,3,-1).reshape([2,3])

**Exercise 4**

In [ ]:
ary_1 = np.ones([3,2])
ary_2 = np.arange(1,7).reshape(3,2)
new_ary = ary_1 + ary_2
new_ary[np.logical_and(new_ary > 2, new_ary < 6)]

**Exercise 5**

In [ ]:
my_mat=np.matrix([[3,2,-1],[2,-2,4],[-1,0.5,-1]])
my_mat.I*np.matrix([1,-2,0]).T

**Exercise 6**

In [ ]:
seto_sep_len=sepal_len[iris_raw['target']==0]
vers_sep_len=sepal_len[iris_raw['target']==1]
virg_sep_len=sepal_len[iris_raw['target']==2]
stats.f_oneway(seto_sep_len,vers_sep_len,virg_sep_len)